build/repo.targets

```diff
@@ -1,56 +1,195 @@
+
+    <Warning Text="KOREBUILD_REPOSITORY_EXCLUDE AND KOREBUILD_REPOSITORY_INCLUDE are specified."
+      Condition="'$(KOREBUILD_REPOSITORY_INCLUDE)' != '' AND '$(KOREBUILD_REPOSITORY_EXCLUDE)' != ''" />
+
+    <Message Text="%(Repository.CloneUrl)" />
   </Target>
 
-  <Target Name="CIBuild" DependsOnTargets="SetBuildNumber">
+  <Target Name="CloneRepositories">
+    <ItemGroup>
+      <_CloneRepository Include="$(MSBuildProjectFullPath)">
+        <AdditionalProperties>
+          CloneRepository=%(Repository.Identity);
+          CloneUrl=%(Repository.CloneUrl);
+          CloneBranch=%(Repository.Branch);
+          CloneRepositoryCommit=%(Repository.Commit);
+          ShallowClone=true;
+          UseGateBranch=$(UseGateBranch)
+        </AdditionalProperties>
+      </_CloneRepository>
+    </ItemGroup>
+
+    <RemoveDir Directories="$(_CloneRepositoryRoot)" Condition="Exists('$(_CloneRepositoryRoot)')" />
+    <MakeDir Directories="$(_CloneRepositoryRoot)" />
+
+    <MSBuild Projects="@(_CloneRepository)"
+      Targets="_CloneRepository"
+      BuildInParallel="$(BuildInParallel)" />
+  </Target>
+
+  <Target Name="_CloneRepository">
     <PropertyGroup>
-      <_SakeTargets>$(_SakeTargets):ci-build</_SakeTargets>
+      <ShallowClone Condition="'$(ShallowClone)'=='' OR '$(CloneRepositoryCommit)'!=''">false</ShallowClone>
+      <_CloneArguments>git clone --quiet $(CloneUrl)</_CloneArguments>
+      <_CloneArguments Condition="'$(ShallowClone)'=='true'">$(_CloneArguments) --depth 1</_CloneArguments>
     </PropertyGroup>
-    <MSBuild Projects="$(MSBuildProjectFullPath)"
-      Targets="Sake"
-      Properties="SakeTargets=$(_SakeTargets);RepositoryRoot=$(RepositoryRoot)" />
+
+    <Message Text="Cloning $(CloneRepository) ..." Importance="High" />
+    <Exec
+      Command="$(_CloneArguments) --branch $(CloneBranch)-gate"
+      Condition="'$(UseGateBranch)'=='true'"
+      IgnoreExitCode="true"
+      IgnoreStandardErrorWarningFormat="true"
+      WorkingDirectory="$(_CloneRepositoryRoot)">
+      <Output TaskParameter="ExitCode" PropertyName="GateBranchExitCode" />
+    </Exec>
+
+    <Warning Text="Using $(CloneBranch)-gate for $(CloneRepository)"
+      Condition="'$(GateBranchExitCode)'=='0'" />
+
+    <Exec
+      Command="$(_CloneArguments) --branch $(CloneBranch)"
+      Condition="'$(GateBranchExitCode)'!='0'"
+      WorkingDirectory="$(_CloneRepositoryRoot)" />
+
+    <Exec
+      Command="git reset --quiet --hard $(CloneRepositoryCommit)"
+      WorkingDirectory="$(_CloneRepositoryRoot)$(CloneRepository)"
+      Condition="'$(CloneRepositoryCommit)'!=''" />
   </Target>
 
-  <Target Name="CITest" DependsOnTargets="SetBuildNumber">
-  <PropertyGroup>
-      <_SakeTargets>$(_SakeTargets):ci-test</_SakeTargets>
-    </PropertyGroup>
-    <MSBuild Projects="$(MSBuildProjectFullPath)"
-      Targets="Sake"
-      Properties="SakeTargets=$(_SakeTargets);RepositoryRoot=$(RepositoryRoot)" />
+  <Target Name="BuildRepositories"
```

Comment:
This always attempts to produce a coherent build from the list of cloned repositories. Doing this includes fiddling around with repo's NuGet.configs + pinning package versions. We could add a different (default?) target that does what `.\build` used to do with sake - i.e build the repositories in sequence (ever since we broke `build install` (https://github.com/aspnet/KoreBuild/issues/17). 

That said, it might be simpler to say this repo is primarily a way for our build system to co-ordinate builds rather than for users to build all of AspNet. I'm not sure there's a lot of value in doing this. Thoughts @davidfowl ?